content/en/docs/components/pipelines/v1/sdk/connect-api.md

```diff
@@ -152,146 +152,186 @@ and give them access to submit pipelines in their own namespace.
+
+        # use a persistent session (for cookies)
+        s = requests.Session()
+
+        # disable SSL verification, if requested
+        if self._skip_tls_verify:
+            s.verify = False
+            urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+        # GET the api_url, which should redirect to Dex
+        resp = s.get(self._api_url, allow_redirects=True)
         if resp.status_code != 200:
             raise RuntimeError(
-                f"HTTP status code '{resp.status_code}' for GET against: {url}"
+                f"HTTP status code '{resp.status_code}' for GET against: {self._api_url}"
             )
 
-        auth_session["redirect_url"] = resp.url
-
-        # if we were NOT redirected, then the endpoint is UNSECURED
+        # if we were NOT redirected, then the endpoint is unsecured
         if len(resp.history) == 0:
-            auth_session["is_secured"] = False
-            return auth_session
-        else:
-            auth_session["is_secured"] = True
-
-        ################
-        # Get Dex Login URL
-        ################
-        redirect_url_obj = urlsplit(auth_session["redirect_url"])
+            # no cookies are needed
+            return ""
 
         # if we are at `/auth?=xxxx` path, we need to select an auth type
-        if re.search(r"/auth$", redirect_url_obj.path): 
-            
-            #######
-            # TIP: choose the default auth type by including ONE of the following
-            #######
-            
-            # OPTION 1: set "staticPasswords" as default auth type
-            redirect_url_obj = redirect_url_obj._replace(
-                path=re.sub(r"/auth$", "/auth/local", redirect_url_obj.path)
+        url_obj = urlsplit(resp.url)
+        if re.search(r"/auth$", url_obj.path):
+            url_obj = url_obj._replace(
+                path=re.sub(r"/auth$", f"/auth/{self._dex_auth_type}", url_obj.path)
             )
-            # OPTION 2: set "ldap" as default auth type 
-            # redirect_url_obj = redirect_url_obj._replace(
-            #     path=re.sub(r"/auth$", "/auth/ldap", redirect_url_obj.path)
-            # )
-            
-        # if we are at `/auth/xxxx/login` path, then no further action is needed (we can use it for login POST)
-        if re.search(r"/auth/.*/login$", redirect_url_obj.path):
-            auth_session["dex_login_url"] = redirect_url_obj.geturl()
-
-        # else, we need to be redirected to the actual login page
+
+        # if we are at `/auth/xxxx/login` path, then we are at the login page
+        if re.search(r"/auth/.*/login$", url_obj.path):
+            dex_login_url = url_obj.geturl()
         else:
-            # this GET should redirect us to the `/auth/xxxx/login` path
-            resp = s.get(redirect_url_obj.geturl(), allow_redirects=True)
+            # otherwise, we need to follow a redirect to the login page
+            resp = s.get(url_obj.geturl(), allow_redirects=True)
```

Comment:
I think `allow_redirects=True` has been used earlier in `resp = s.get(self._api_url, allow_redirects=True)`, in what situation do we need to manually redirect here? Can we use an code `301` from response to determine it is a `redirect` response?

Reply:
The problem is that multiple redirects can happen (and we don't know how many, because it depends on the user's dex configs, specifically, if they have multiple auth providers, or just one).

So using  `allow_redirects` makes it cleaner.

Reply:
I am not against using `allow_redirects=True`, but the `else` condition here is trying to call `resp = s.get(url_obj.geturl(), allow_redirects=True)` again, while `resp = s.get(self._api_url, allow_redirects=True)` earlier should have handled the redirect automatically.

If multiple manual `s.get()` is needed, it sounds like a `for loop` is needed, instead of `if/else`.

Reply:
@zijianjoy actually, I think we are slightly misunderstanding what's happening here (its been a while since I wrote this).

First, we do a get which may redirect us to a "choose auth type" page, if we land on that page, the following code rewrites the patch to the "chosen" auth type:

```python
        # if we are at `/auth?=xxxx` path, we need to select an auth type
        url_obj = urlsplit(resp.url)
        if re.search(r"/auth$", url_obj.path):
            url_obj = url_obj._replace(
                path=re.sub(r"/auth$", f"/auth/{self._dex_auth_type}", url_obj.path)
            )
```

After this, we need to perform another get to be redirected to the actual login page.

Reply:
Thanks for the explanation, in that case, shouldn't we make `get call` immediately after this block?

```
  # if we are at `/auth?=xxxx` path, we need to select an auth type
        url_obj = urlsplit(resp.url)
        if re.search(r"/auth$", url_obj.path):
            url_obj = url_obj._replace(
                path=re.sub(r"/auth$", f"/auth/{self._dex_auth_type}", url_obj.path)
            )
            resp = s.get(url_obj.geturl(), allow_redirects=True)
```